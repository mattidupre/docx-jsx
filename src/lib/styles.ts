import { kebabCase, pick } from 'lodash';
import {
  TYPOGRAPHY_CSS_KEYS,
  type Variants,
  INTRINSIC_VARIANT_TAG_NAMES,
  type PrefixesConfig,
  type TypographyOptions,
  type VariantName,
  type FontsConfig,
} from '../entities';
import { getValueOf } from '../utils/object';
import { joinKebab, prefixKebab } from '../utils/string';
import {
  type CssRulesArray,
  cssRulesArrayToString,
  type CssRuleDeclarations,
  objectToVarValues,
} from '../utils/css';

export const variantNameToClassName = (
  { prefixes }: { prefixes: PrefixesConfig },
  variantName: VariantName,
) => prefixKebab(prefixes.variantClassName, variantName);

export const typographyOptionsToStyleVars = (
  options: {
    prefixes: Pick<PrefixesConfig, 'cssVariable'>;
  },
  typographyOptions: undefined | TypographyOptions,
) =>
  objectToVarValues(pick(typographyOptions, ...TYPOGRAPHY_CSS_KEYS), {
    prefix: options.prefixes.cssVariable,
  });

export const variantsToStyleVars = (
  options: {
    prefixes: Pick<PrefixesConfig, 'cssVariable'>;
  },
  variants: Partial<Variants>,
) =>
  objectToVarValues(variants, {
    prefix: options.prefixes.cssVariable,
  });

const createPrefixedVar = (
  options: {
    prefixes: Pick<PrefixesConfig, 'cssVariable'>;
  },
  value: string | number,
  defaultValue?: string | number,
) =>
  `var(--${joinKebab(options.prefixes.cssVariable, String(value))}${
    defaultValue !== undefined ? `, ${defaultValue}` : ''
  })`;

const DEFAULT_VARS: CssRuleDeclarations = {
  fontSize: '1rem',
  borderWidth: 0,
  borderTopWidth: 0,
  borderRightWidth: 0,
  borderBottomWidth: 0,
  borderLeftWidth: 0,
};

export const createStyleArray = (options: {
  variants?: Variants;
  prefixes: Pick<PrefixesConfig, 'cssVariable' | 'variantClassName'>;
  fonts?: FontsConfig;
}): CssRulesArray => {
  const { prefixes, variants = {}, fonts = [] } = options;

  // TODO: Add @font-face rules.

  // TODO: How to handle highlight, <sup>, <sub>?
  const rules: CssRulesArray = [];

  rules.push([
    '*',
    TYPOGRAPHY_CSS_KEYS.reduce(
      (declarations, property) => {
        if (property === 'marginTop' || property === 'marginBottom') {
          return declarations;
        }
        return {
          ...declarations,
          [kebabCase(property)]: createPrefixedVar(
            options,
            property,
            DEFAULT_VARS[property] ?? 'unset',
          ),
        };
      },
      {
        borderWidth: 0,
        borderStyle: 'solid',
      } as CssRuleDeclarations,
    ),
  ]);

  rules.push([
    `:where(${['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p', 'ul', 'ol', 'dl'].join(
      ', ',
    )})`,
    {
      marginTop: createPrefixedVar(options, 'marginTop', 0),
      marginBottom: createPrefixedVar(options, 'marginBottom', 0),
    },
  ]);

  // Variant vars are attached to variant class names .[prefix]-[variant-name]
  for (const variantName in variants) {
    const intrinsicTagName = getValueOf(
      INTRINSIC_VARIANT_TAG_NAMES,
      variantName,
    );

    const variantSelectors = [
      `:where(.${variantNameToClassName({ prefixes }, variantName)})`,
      intrinsicTagName ? `:where(${intrinsicTagName})` : [],
    ].flat();

    const variantSelectorSelf = variantSelectors.join(', ');

    rules.push([
      // Variant tags set variant CSS vars.
      variantSelectorSelf,
      typographyOptionsToStyleVars({ prefixes }, variants[variantName]),
    ]);
  }

  return rules;
};

export const createStyleString = (
  ...args: Parameters<typeof createStyleArray>
) => cssRulesArrayToString(createStyleArray(...args));
