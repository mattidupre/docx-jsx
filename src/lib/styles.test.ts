import { test, expect } from 'vitest';
import {
  createMockPrefixesConfig,
  createMockVariantsConfig,
} from '../fixtures';
import { createStyleArray, createStyleString } from './styles';

const createOptions = () => ({
  variants: createMockVariantsConfig(),
  prefixes: createMockPrefixesConfig(),
});

test('createStyleArray', () => {
  const cssRules = createStyleArray(createOptions());
  expect(cssRules).toMatchSnapshot();

  for (const rule of cssRules) {
    console.log(`${rule[0]}: \n${JSON.stringify(rule[1], undefined, 2)}\n`);
  }
});

test('createStyleString', () => {
  const cssString = createStyleString(createOptions());
  console.log(cssString);
});
