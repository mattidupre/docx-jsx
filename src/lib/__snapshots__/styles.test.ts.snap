// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createStyleArray 1`] = `
[
  [
    "*",
    {
      "border-bottom-color": "var(--mock-variable-border-bottom-color, unset)",
      "border-bottom-width": "var(--mock-variable-border-bottom-width, 0)",
      "borderStyle": "solid",
      "borderWidth": 0,
      "break-inside": "var(--mock-variable-break-inside, unset)",
      "color": "var(--mock-variable-color, unset)",
      "font-family": "var(--mock-variable-font-family, unset)",
      "font-size": "var(--mock-variable-font-size, 1rem)",
      "font-style": "var(--mock-variable-font-style, unset)",
      "font-weight": "var(--mock-variable-font-weight, unset)",
      "line-height": "var(--mock-variable-line-height, unset)",
      "padding-bottom": "var(--mock-variable-padding-bottom, unset)",
      "text-align": "var(--mock-variable-text-align, unset)",
      "text-decoration": "var(--mock-variable-text-decoration, unset)",
      "text-indent": "var(--mock-variable-text-indent, unset)",
      "text-transform": "var(--mock-variable-text-transform, unset)",
    },
  ],
  [
    ":where(p, h1, h2, h3, h4, h5, h6, li)",
    {
      "display": "flow-root",
      "marginBottom": "var(--mock-variable-margin-bottom, 0)",
      "marginLeft": "var(--mock-variable-margin-left, 0)",
      "marginTop": "var(--mock-variable-margin-top, 0)",
    },
  ],
  [
    ":where(li)",
    {
      "display": "list-item",
    },
  ],
  [
    ":where(.mock-class-name-heading-1), :where(h1)",
    {
      "--mock-variable-color": "#ff00ff",
      "--mock-variable-font-family": "Merriweather",
      "--mock-variable-font-size": "2rem",
      "--mock-variable-font-style": "italic",
      "--mock-variable-font-weight": "bold",
    },
  ],
  [
    ":where(.mock-class-name-hyperlink), :where(a)",
    {
      "--mock-variable-color": "#ff00ff",
    },
  ],
  [
    ":where(.mock-class-name-mock-paragraph-variant)",
    {
      "--mock-variable-border-bottom-color": "#ffff00",
      "--mock-variable-border-bottom-width": "0.25rem",
      "--mock-variable-color": "var(--mock-color, #00dddd)",
      "--mock-variable-font-family": "Merriweather",
      "--mock-variable-font-size": "1.25rem",
      "--mock-variable-font-weight": "bold",
      "--mock-variable-padding-bottom": "0.25rem",
    },
  ],
  [
    ":where(.mock-class-name-mock-text-variant)",
    {
      "--mock-variable-color": "var(--mock-color, #00dddd)",
      "--mock-variable-font-family": "Merriweather",
      "--mock-variable-font-size": "1.25rem",
      "--mock-variable-font-weight": "bold",
    },
  ],
  [
    ":where(.mock-class-name-mock-content-variant)",
    {
      "--mock-variable-line-height": "1.5rem",
      "--mock-variable-margin-bottom": "1rem",
      "--mock-variable-text-indent": "1rem",
    },
  ],
]
`;

exports[`createStyleString 1`] = `
"* {
  border-width: 0;
  border-style: solid;
  break-inside: var(--mock-variable-break-inside, unset);
  text-align: var(--mock-variable-text-align, unset);
  line-height: var(--mock-variable-line-height, unset);
  font-weight: var(--mock-variable-font-weight, unset);
  font-style: var(--mock-variable-font-style, unset);
  font-size: var(--mock-variable-font-size, 1rem);
  font-family: var(--mock-variable-font-family, unset);
  color: var(--mock-variable-color, unset);
  text-transform: var(--mock-variable-text-transform, unset);
  text-decoration: var(--mock-variable-text-decoration, unset);
  text-indent: var(--mock-variable-text-indent, unset);
  padding-bottom: var(--mock-variable-padding-bottom, unset);
  border-bottom-width: var(--mock-variable-border-bottom-width, 0);
  border-bottom-color: var(--mock-variable-border-bottom-color, unset);
}
:where(p, h1, h2, h3, h4, h5, h6, li) {
  margin-top: var(--mock-variable-margin-top, 0);
  margin-bottom: var(--mock-variable-margin-bottom, 0);
  margin-left: var(--mock-variable-margin-left, 0);
  display: flow-root;
}
:where(li) {
  display: list-item;
}
:where(.mock-class-name-heading-1), :where(h1) {
  --mock-variable-font-weight: bold;
  --mock-variable-font-style: italic;
  --mock-variable-font-size: 2rem;
  --mock-variable-font-family: Merriweather;
  --mock-variable-color: #ff00ff;
}
:where(.mock-class-name-hyperlink), :where(a) {
  --mock-variable-color: #ff00ff;
}
:where(.mock-class-name-mock-paragraph-variant) {
  --mock-variable-font-weight: bold;
  --mock-variable-font-size: 1.25rem;
  --mock-variable-font-family: Merriweather;
  --mock-variable-color: var(--mock-color, #00dddd);
  --mock-variable-padding-bottom: 0.25rem;
  --mock-variable-border-bottom-width: 0.25rem;
  --mock-variable-border-bottom-color: #ffff00;
}
:where(.mock-class-name-mock-text-variant) {
  --mock-variable-font-weight: bold;
  --mock-variable-font-size: 1.25rem;
  --mock-variable-font-family: Merriweather;
  --mock-variable-color: var(--mock-color, #00dddd);
}
:where(.mock-class-name-mock-content-variant) {
  --mock-variable-line-height: 1.5rem;
  --mock-variable-text-indent: 1rem;
  --mock-variable-margin-bottom: 1rem;
}"
`;
